base/multidimensional.jl

```diff
@@ -358,8 +358,12 @@ module IteratorsMD
         CartesianIndex()
     end
     @propagate_inbounds function Base.getindex(iter::CartesianIndices{N,R}, I::Vararg{Int, N}) where {N,R}
-        CartesianIndex(getindex.(iter.indices, I))
+        CartesianIndex(_get_cartesianindex.(iter.indices, I))
     end
+    # specialize to gives better hint to compiler to enable SIMD (#42115)
+    @inline _get_cartesianindex(::OneTo, i::Int) = i
+    @inline _get_cartesianindex(::Base.IdentityUnitRange, i::Int) = i
+    @inline _get_cartesianindex(r, i) = getindex(r, i)
```

Comment:
A perhaps more appropriate place to make modifications is adding `@propagate_inbounds` in https://github.com/JuliaLang/julia/blob/876df7929fd96de2975bf172f6034ba3e4e1b73e/base/range.jl#L879-L914

Reply:
The root issue here is that `throw_boundserror(v, i)` is marked as `@noinline` and thus disables the propagation of `@inbounds`

https://github.com/JuliaLang/julia/blob/8812c5cc242435bf267285fcf3e41e632c84e52c/base/abstractarray.jl#L691

Reply:
I don't understand why an error path would do this. But shouldn't the whole expression be in the `@boundscheck`? That is, instead of
```julia
@boundscheck ((i > 0) & ok) || throw_boundserror(v, i) 
```
you just need an extra set of parentheses:
```
@boundscheck(((i > 0) & ok) || throw_boundserror(v, i))
```

Reply:
You're right, this `@boundscheck` does get eliminated if we check individual `getindex(::OneTo, i)`, but it doesn't get eliminated if we check `getindex(::CartesianIndices, I...)`

```julia
getindex_inbounds(r, i...) = @inbounds getindex(r, i...)
R = CartesianIndices((2, 3))
```

```julia
julia> @code_llvm getindex_inbounds(R.indices[1], 1)
;  @ REPL[18]:1 within `getindex_inbounds`
define i64 @julia_getindex_inbounds_2636([1 x i64]* nocapture nonnull readonly align 8 dereferenceable(8) %0, i64 signext %1) #0 {
top:
  ret i64 %1
}

julia> @code_llvm getindex_inbounds(R, 1, 1)
...

L12:                                              ; preds = %top
; │││││└└└
; │││││┌ @ broadcast.jl:642 within `_broadcast_getindex`
; ││││││┌ @ broadcast.jl:666 within `_getindex`
; │││││││┌ @ broadcast.jl:621 within `_broadcast_getindex`
; ││││││││┌ @ tuple.jl:29 within `getindex`
           %9 = getelementptr inbounds [1 x [2 x [1 x i64]]], [1 x [2 x [1 x i64]]]* %1, i64 0, i64 0, i64 0
; │││││└└└└
; │││││┌ @ broadcast.jl:643 within `_broadcast_getindex`
; ││││││┌ @ broadcast.jl:670 within `_broadcast_getindex_evalf`
; │││││││┌ @ range.jl:901 within `getindex`
          %10 = call nonnull {}* @j_throw_boundserror_2682([1 x i64]* nocapture nonnull readonly %9, i64 signext %2) #0
          call void @llvm.trap()
          unreachable
...

L26:                                              ; preds = %L18
          %17 = call nonnull {}* @j_throw_boundserror_2681([1 x i64]* nocapture readonly %11, i64 signext %3) #0
          call void @llvm.trap()
          unreachable
...
```

Notice the two `throw_boundserror` call. Thus I did an eager check so that we can pass `@inbounds` hint through the broadcasting/map. After 20c4be5:

```julia
julia> @code_llvm getindex_inbounds(R, 1, 1)
;  @ REPL[18]:1 within `getindex_inbounds`
define void @julia_getindex_inbounds_2685([1 x [2 x i64]]* noalias nocapture sret([1 x [2 x i64]]) %0, [1 x [2 x [1 x i64]]]* nocapture nonnull readonly align 8 dereferenceable(16) %1, i64 signext %2, i64 signext %3) #0 {
top:
  %.sroa.0.sroa.0.0..sroa.0.0..sroa_cast4.sroa_idx = getelementptr inbounds [1 x [2 x i64]], [1 x [2 x i64]]* %0, i64 0, i64 0, i64 0
  store i64 %2, i64* %.sroa.0.sroa.0.0..sroa.0.0..sroa_cast4.sroa_idx, align 8
  %.sroa.0.sroa.2.0..sroa.0.0..sroa_cast4.sroa_idx8 = getelementptr inbounds [1 x [2 x i64]], [1 x [2 x i64]]* %0, i64 0, i64 0, i64 1
  store i64 %3, i64* %.sroa.0.sroa.2.0..sroa.0.0..sroa_cast4.sroa_idx8, align 8
  ret void
}
```